lib/resque/plugins/job_history/history_list.rb

```diff
@@ -86,11 +86,13 @@ def max_jobs
         end
 
         def delete_old_jobs(job_count)
-          while job_count > max_jobs
-            job_id = redis.rpop(job_list_key)
+          while num_jobs > max_jobs
+            job_id         = redis.rpop(job_list_key)
+            job_class_name = job_class(job_id)
```

Comment:
In the expected/common case of a history for a class, this just returns the `class_name`.

For linear histories, this checks Redis for the the `class_name` for the `job_id`.

The old code deleted the `job_id` from the `class_name` list before trying to look it up, resulting in the system not knowing which class contains the passed in `job_id` and therefore preventing it from being deleted.